pkg/kvstore/store/store.go

```diff
@@ -0,0 +1,429 @@
+
+	if err := s.listAndStartWatcher(); err != nil {
+		return nil, err
+	}
+
+	controllers.UpdateController(s.controllerName,
+		controller.ControllerParams{
+			DoFunc: func() error {
+				return s.syncLocalKeys()
+			},
+			RunInterval: s.conf.SynchronizationInterval,
+		},
+	)
+
+	return s, nil
+}
+
+// Close stops participation with a shared store. This stops the controller
+// started by JoinSharedStore().
+func (s *SharedStore) Close() {
+	// Wait for all write operations to complete and then block all further
+	// operations
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	if s.kvstoreWatcher != nil {
+		s.kvstoreWatcher.Stop()
+	}
+
+	controllers.RemoveController(s.controllerName)
+
+	for name, key := range s.localKeys {
+		if err := kvstore.Delete(s.keyPath(key)); err != nil {
+			s.getLogger().WithError(err).Warning("Unable to delete key in kvstore")
+		}
+
+		delete(s.localKeys, name)
+		key.OnDelete()
+	}
+}
+
+// keyPath returns the absolute kvstore path of a key
+func (s *SharedStore) keyPath(key Key) string {
+	// WARNING - STABLE API: The composition of the absolute key path
+	// cannot be changed without breaking up and downgrades.
+	return path.Join(s.conf.Prefix, key.GetKeyName())
+}
+
+// syncLocalKey synchronizes a key to the kvstore
+func (s *SharedStore) syncLocalKey(key LocalKey) error {
+	jsonValue, err := key.Marshal()
+	if err != nil {
+		return err
+	}
+
+	// Update key in kvstore, overwrite an eventual existing key, attach
+	// lease to expire entry when agent dies and never comes back up.
+	if err := kvstore.Update(s.keyPath(key), jsonValue, true); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// syncLocalKeys synchronizes all local keys with the kvstore
+func (s *SharedStore) syncLocalKeys() error {
+	// Create a copy of all local keys so we can unlock and sync to kvstore
+	// without holding the lock
+	s.mutex.RLock()
+	keys := []LocalKey{}
+	for _, key := range s.localKeys {
+		keys = append(keys, key)
+	}
+	s.mutex.RUnlock()
+
+	for _, key := range keys {
+		if err := s.syncLocalKey(key); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// UpdateLocalKey adds a key to be synchronized with the kvstore
+func (s *SharedStore) UpdateLocalKey(key LocalKey) {
+	s.mutex.Lock()
+	s.localKeys[key.GetKeyName()] = key
+	s.mutex.Unlock()
+}
+
+// UpdateLocalKeySync synchronously synchronizes a local key with the kvstore
+// and adds it to the list of local keys to be synchronized if the initial
+// synchronous synchronization was successful
+func (s *SharedStore) UpdateLocalKeySync(key LocalKey) error {
+	s.UpdateLocalKey(key)
+
+	if err := s.syncLocalKey(key); err != nil {
+		s.DeleteLocalKey(key)
+		return err
+	}
+
+	return nil
+}
+
+// DeleteLocalKey removes a key from bein synchronized with the kvstore
+func (s *SharedStore) DeleteLocalKey(key LocalKey) {
+	err := kvstore.Delete(s.keyPath(key))
+	name := key.GetKeyName()
+
+	s.mutex.Lock()
+	delete(s.localKeys, name)
```

Comment:
if you `delete(s.localKeys, name)` and then check `_, ok := s.localKeys[name]`, won't `ok` always be `false`?